### PR TITLE
Liq 2.0

### PR DESCRIPTION
**Pull request by gbalabasquer**
Final PR for merging liquidations 2.0 to `master`.

### CODE REVIEW

File: src/clip.sol [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+        uint256 lot,
+        address indexed usr,
+        address indexed kpr,
+        uint256 coin
+    );
+
+    event Yank(uint256 id);
+
+    // --- Init ---
+    constructor(address vat_, address spotter_, address dog_, bytes32 ilk_) public {
```
Comment (hexonaut) [inline src/clip.sol:135]: Constructor visibility is redundant. It is assumed to be public and can be safely removed.

File: src/clip.sol [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+
+    // The number of active auctions
+    function count() external view returns (uint256) {
+        return active.length;
+    }
+
+    // Return the entire array of active auctions
+    function list() external view returns (uint256[] memory) {
+        return active;
+    }
```
Comment (hexonaut) [inline src/clip.sol:436]: `active` has a `public` visibility modifier. This seems redundant. Why is it provided?

File: src/clip.sol [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+
+    // Cancel an auction during ES or via governance action.
+    function yank(uint256 id) external auth lock {
+        require(sales[id].usr != address(0), "Clipper/not-running-auction");
+        dog.digs(ilk, sales[id].tab);
+        vat.flux(ilk, address(this), msg.sender, sales[id].lot);
+        _remove(id);
+        emit Yank(id);
+    }
```
Comment (hexonaut) [inline src/clip.sol:471]: I think we should start including this on every contract to reduce the need for this `upchost`-style caching with big structs:

```
function _peek(bytes32 addr) external pure returns (bytes32 b) {
    assembly {
        b := mload(addr)
    }
}
```

**Review by hexonaut — COMMENTED** [review]
Mostly style changes, but also recommending we start adding `_peek()` for arbitrary storage lookups in all core contracts to avoid this `chost`-style caching in the future.

File: src/abaci.sol [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    uint256 public tau;  // Seconds after auction start when the price reaches zero [seconds]
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+
+    event FileUint256(bytes32 indexed what, uint256 data);
+
+    // --- Init ---
+    constructor() public {
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/abaci.sol:48]: No, it is actually not for solc 0.6.12 or lower. I think this applies for >= 0.7.

```
dog.sol:102:5: SyntaxError: No visibility specified. Did you intend to add "public"?
```

File: src/abaci.sol [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    // --- Data ---
+    uint256 public tau;  // Seconds after auction start when the price reaches zero [seconds]
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+
+    event FileUint256(bytes32 indexed what, uint256 data);
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/abaci.sol:45]: We forgot to update the events in this file. Great catch.

File: src/dog.sol [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        if (what == "clip") {
+            require(ilk == ClipperLike(clip).ilk(), "Dog/file-ilk-neq-clip.ilk");
+            ilks[ilk].clip = clip;
+        } else revert("Dog/file-unrecognized-param");
+        emit File(ilk, what, clip);
+    }
+
+    function chop(bytes32 ilk) external view returns (uint256) {
+        return ilks[ilk].chop;
+    }
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/dog.sol:154]: This is a good one to discuss for liquidation 2.0

File: src/dog.sol [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            require(spot > 0 && mul(ink, spot) < mul(art, rate), "Dog/not-unsafe");
+
+            // Get the minimum value between:
+            // 1) Remaining space in the general Hole
+            // 2) Remaining space in the collateral hole
+            require(Hole > Dirt && milk.hole > milk.dirt, "Dog/liquidation-limit-hit");
+            uint256 room = min(Hole - Dirt, milk.hole - milk.dirt);
+
+            // uint256.max()/(RAD*WAD) = 115,792,089,237,316
+            dart = min(art, mul(room, WAD) / rate / milk.chop);
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/dog.sol:190]: I guess we prioritize the exactness over the highest quantity, which as you mention is already quite high.

File: src/clip.sol (hunk 2 above)
Comment (gbalabasquer) [reply src/clip.sol:436]: No, `active()` expects an index and returns the value of that index. `list()` returns the whole array.

File: src/clip.sol (hunk 3 above)
Comment (gbalabasquer) [reply src/clip.sol:471]: As I said in a prev comment, we should discuss this for liquidations 2.1. Maybe it is just easier to add all the methods manually when we have a struct, but this is also a cool idea that as you mentioned works in the meanwhile we do not want to protect an onchain read for a value.

File: src/abaci.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/abaci.sol:48]: Ah okay. My mistake.

**Review by godsflaw — APPROVED** [review]
LGTM

**Review by e18r — APPROVED** [review]
LGTM

File: src/dog.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (godsflaw) [reply src/dog.sol:154]: There are pros and cons to this approach, we should discuss for the next revision.

**Review by WilfredTA — APPROVED** [review]
lgtm

File: src/dog.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (kmbarry1) [reply src/dog.sol:154]: Yeah I've considered this, might be worth doing. Here, `chop()` was specifically added by request for B.Protocol, so it predates this realization of the need to gas optimize storage reads.

[the remark this answers is not in the document]
Comment (kmbarry1) [reply src/dog.sol:154]: (and, I should add, it's much friendlier for integrators this way as they don't need to rely on a storage slot which could have its use changed in an update, or even the parameter could be redefined or expressed differently)

File: src/clip.sol [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+    function wmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = mul(x, y) / WAD;
+    }
+    function rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = mul(x, y) / RAY;
+    }
+    function rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = mul(x, RAY) / y;
+    }
+
```
Comment (kmbarry1) [inline src/clip.sol:204]:
```suggestion
```

**Review by kmbarry1 — APPROVED** [review]
LGTM

**Pull request merged**